src/allocator.rs

```diff
@@ -0,0 +1,232 @@
+// TODO(i.erin) iterator over buffer chain?
+
+/// Intrusive linked list of buffers.
+#[repr(C)]
+struct BufferLayout {
+    // Pointer to the next entry of the intrusive linked list.
+    next: Option<Buffer>,
+
+    // Allocation layout.
+    layout: Layout,
+
+    // Actuall data.
+    payload: [u8],
+}
+
+struct Buffer(NonNull<BufferLayout>);
+
+impl Buffer {
+    const HEADER_BUFFER_FIELDS: &[Layout] =
+        &[Layout::new::<Option<Buffer>>(), Layout::new::<Layout>(), Layout::new::<usize>()];
+
+    pub fn alloc(size: usize) -> Buffer {
+        // TODO(i.erin)
+        let layout = Self::compute_layout(size).unwrap();
+
+        // Safety: non zero layout size.
+        assert!(layout.size() != 0);
+        let ptr = unsafe { alloc::alloc_zeroed(layout) };
+        if ptr.is_null() {
+            alloc::handle_alloc_error(layout)
+        }
+
+        // Safety: non null, memory of size `size`, owning pointer
+        // Discassed: https://users.rust-lang.org/t/can-i-create-a-reference-to-a-custom-dst-from-raw-parts-on-stable/63261
+        assert!(!ptr.is_null());
+        let ptr: *mut [()] = unsafe { std::slice::from_raw_parts_mut(ptr as *mut (), size) };
+
+        // Rustonumicon: https://doc.rust-lang.org/stable/reference/expressions/operator-expr.html#pointer-to-pointer-cast
+        let ptr = ptr as *mut BufferLayout;
+
+        unsafe { (*ptr).next = None };
+        unsafe { (*ptr).layout = layout };
+        // payload already initialazed via alloca_zeroed
+
+        // Safety: nullness checked right after allocation
+        unsafe { Buffer(NonNull::new_unchecked(ptr)) }
+    }
+
+    // Safety: should be called only once for pointer created via [`Self::alloc`].
+    pub unsafe fn dealloc(&mut self) {
+        let ptr = self.0.as_ptr();
+
+        let layout = unsafe { (*ptr).layout };
+        unsafe { alloc::dealloc(ptr as *mut _, layout) };
+    }
+
+    pub fn as_mut(&mut self) -> &mut [u8] {
+        self.as_mut_and_next().0
+    }
+
+    pub fn mut_next(&mut self) -> &mut Option<Buffer> {
+        self.as_mut_and_next().1
+    }
+
+    pub fn as_mut_and_next(&mut self) -> (&mut [u8], &mut Option<Buffer>) {
+        // Safety: TODO()
+        let as_mut = unsafe { &mut (*self.0.as_ptr()).payload };
+        let next = unsafe { &mut (*self.0.as_ptr()).next };
+
+        (as_mut, next)
+    }
+
+    fn repr_c_layout(fields: &[Layout]) -> Result<Layout, LayoutError> {
+        let mut layout = Layout::from_size_align(0, 1)?;
+        for &field in fields {
+            let (new_layout, _) = layout.extend(field)?;
+            layout = new_layout;
+        }
+        Ok(layout.pad_to_align())
+    }
+
+    fn compute_layout(layout_size: usize) -> Result<Layout, LayoutError> {
+        let header_layout = Self::repr_c_layout(Self::HEADER_BUFFER_FIELDS)?;
+        let (layout, _) =
+            header_layout.extend(Layout::for_value(vec![0u8; layout_size].as_slice()))?;
+
+        Ok(layout.pad_to_align())
+    }
```

Comment:
use array instead of vector

```suggestion
    fn compute_layout(payload_size: usize) -> Result<Layout, LayoutError> {
        let header_layout = Self::repr_c_layout(Self::HEADER_BUFFER_FIELDS)?;

        let payload_layout = Layout::array::<u8>(payload_size)?;
        let (full_layout, _) = header_layout.extend(payload_layout)?;

        Ok(full_layout.pad_to_align())
    }
```

Reply:
nice!